cmd/autoheal/healer.go

```diff
@@ -133,6 +156,10 @@ func (h *Healer) Run(stopCh <-chan struct{}) error {
 	// Start the workers:
 	go wait.Until(h.runRulesWorker, time.Second, stopCh)
 	go wait.Until(h.runAlertsWorker, time.Second, stopCh)
+	// check for expired cells every interval.
+	if !h.throttling_disabled {
```

Comment:
I think that the fact that expired cells need to be purged should be hidden to the user of the memory object. If you need to do some cleanup when stopping then I'd suggest to add a `Close` method to the `Healer` type. That method should in turn call the `Close` method of the memory, which should then stop the goroutine that does the purging.

Note also that purging at intervals is a poor solution when you know exactly when you need to do the next purging. However it is acceptable in this case because it is combined with purging each time you call the `Has` method.

Reply:
We can also lose the purging worker altogether if you think it's okay to keep action items in memory in between `Has` calls. Note that next purging is *per action* - if my interval is 1h and I added three actions to actionMemory: one at 10m one at 20m and one at 30m than purging the first action should only be valid from 70m onwards, the second at 80m onwards and the third and 90m onwards.. So there is no *exact* purge - the purge time is different between different actions. But this seems like the most simple solution to me.. Do you have any suggestions other than running purge periodically paired with purge on `Has`?

Reply:
I don't see a big problem with keeping action instances in memory. In the worst case it would waste some memory till the next action is triggered.

Exact purge is important, but we already have it because we do i tin the `Has` method. I think we don't need more than that.

If we wanted to implement exact purge without doing it in the `Has` method we could do the following:

1. Have a `nextPurge` field in the type. This should contain the exact time when the next purge needs to be done.

2. Have a `purgeCh` channel.

3. Have a gorutine that creates a timer with the value of the `nextPurge` field, and then waits for the timer to expire or for the next message in the `purgeCh` channel. The message in the channel can be anything, for example `interface{}`, it is just a mechanism to wake up that goroutine. When the message in the channel is `nil` the goroutine finishes.

4. In the methods that update the memory also update the `nextPurge` value, and send an empty (but not `nil`) message to the `purgeCn` channel.

5. When the memory is closed, send `nil` to the `purge` channel.

With that we would only purge exactly when it is needed. But this is probably overkill, I think we don't really need it, your current solution is enough.